tests/e2e/e2e_test.go

```diff
@@ -62,37 +64,20 @@ var _ = Describe("EBS CSI Driver", func() {
 
 		volume := resp.GetVolume()
 		Expect(volume).NotTo(BeNil(), "Expected valid volume, got nil")
-
-		// Verifying that volume was created ans is valid
-		descParams := &ec2.DescribeVolumesInput{
-			Filters: []*ec2.Filter{
-				{
-					Name:   aws.String("tag:" + cloud.VolumeNameTagKey),
-					Values: []*string{aws.String(req.GetName())},
-				},
-			},
-		}
-		waitForVolumes(descParams, 1 /* number of expected volumes */)
+		waitForVolumeState(volume.Id, "available")
```

Comment:
Feels to me its better to move the `watiForVolume` creation and deletion inside driver `CreateVolume` and `DeleteVolume` so that the volume is guaranteed to be available/removed after the call completes. This also void the case where `ControllerPublishVolume` is called during volume is still being created. 

We could do similar thing for attach (saw your TODO comment on `AttachDisk`) and detach.

Reply:
I agree that we need to have similar functionality in the `cloud` package (e.g., when creating, deleting, attaching and detaching volumes). However, it seems like the implementation of these functions in the driver should be a lot more sophisticated than this. For instance, the function that waits for the volume attachment status in Kubernetes has some magic [1] that I don't quite understand yet and can be tricky to write (in the right way).

Also, assuming that `CreateDisk` only returns when the volume is ready, e.g., with `available` status, I think it's the task of the test case to make sure that the volume is indeed created here.

[1] https://github.com/bertinatto/kubernetes/blob/5fdb0c7177010711ca7b463c7e1d4f32a163a344/pkg/cloudprovider/providers/aws/aws.go#L1862

Reply:
Got ya. We can punt this to #38 when we have time to properly implement this.